File: 0491. Non-decreasing Subsequences/Solution0491.py
class Solution0491:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        def dfs(i, prev, curr):
            if len(curr) >= 2:
                ans.add(curr[:])
            if i >= len(nums):
                return
            for j in range(i, len(nums)):
                if nums[j] >= prev:
                    dfs(j+1, nums[j], curr+(nums[j],))
        ans = set()
        dfs(0, -float('inf'), ())
        return ans

File: 0491. Non-decreasing Subsequences/Solution0491.cpp
class Solution0491 {
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> curr;
        dfs(res, 0,  nums,curr);
        return res;
    }

    void dfs(vector<vector<int>>& res, int idx, vector<int>& nums , vector<int>& curr) {
        if (idx >= nums.size()) {
            if (curr.size() >= 2) {
                res.push_back(curr);
            }          
            return;
        }
        if (curr.size() == 0 || nums[idx] >= curr[curr.size()-1]) {
            curr.push_back(nums[idx]);
            dfs(res, idx+1, nums, curr);
            curr.pop_back();
        }
        if (idx > 0 && curr.size() > 0 && nums[idx]==curr[curr.size()-1]) {
            return;
        }
        dfs(res, idx+1, nums, curr);
    }
};